### PR TITLE
Adds `odo dev --no-watch`

### LINKED ISSUE #5631: `odo dev --no-watch`
## User Story
- As an odo user
- I want to be able to run `odo dev` command without triggering push with when files change
- So that the sync and application restart is not constantly happening when I'm making a lot of changes in my ide/editor.

note: There is a followup story that will allow users to trigger push "on-demand" https://github.com/redhat-developer/odo/issues/5634

## Acceptance Criteria
- [ ] `odo dev` command should have `--no-watch` flag. When this flag is specified the command should not watch files for changes.
- [ ] the push happens only once when the command is executed

/kind user-story

### PR DESCRIPTION
**Pull request by dharmit**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**
Need this to allow users to avoid syncing the code too fast

**Which issue(s) this PR fixes:**
Fixes #5631

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
1. Execute `odo dev --no-watch`; modify a file in the component directory.
2. odo should not trigger a sync for the file change
3. Hitting `Ctrl+c` should correctly delete the resources from the cluster

### CODE REVIEW

**Comment by dharmit** [comment]
/test unit

Looks like network issue
```
 go mod verify
go: github.com/Netflix/go-expect@v0.0.0-20201125194554-85d881c3777e: Get "https://proxy.golang.org/github.com/%21netflix/go-expect/@v/v0.0.0-20201125194554-85d881c3777e.mod": dial tcp: i/o timeout
make[1]: *** [check-vendor] Error 1
make[1]: Leaving directory `/go/src/github.com/redhat-developer/odo'
make: *** [openshiftci-presubmit-unittests] Error 2 
```

File: pkg/watch/interface.go
Comment (valaparthvi) [inline pkg/watch/interface.go:15]: You'll need to update the mock client after this change.

**Review by valaparthvi — CHANGES_REQUESTED** [review]

File: pkg/watch/interface.go
Comment (dharmit) [reply pkg/watch/interface.go:15]: 7b82c4a Apparently there were other mocks as well that got updated. 😂

File: tests/integration/devfile/cmd_dev_test.go
Comment (rm3l) [inline tests/integration/devfile/cmd_dev_test.go:351]: Do we need a call to `filepath.Base` here? I guess it would make sense if we were manipulating random files, but here we already know what the file name is.

```suggestion
					execResult := commonVar.CliRunner.Exec(podName, commonVar.Project, "cat", "/projects/server.js")
```

File: pkg/watch/interface.go
Comment (rm3l) [reply pkg/watch/interface.go:15]: Isn't the `Func` suffix redundant? This is already a method. :)
I think the `Func` suffix would have made sense if this was returning a function, but this is not the case here.

**Review by rm3l — CHANGES_REQUESTED** [review]

**Review by rm3l — COMMENTED** [review]
One last thing I noticed after testing your branch. The color of the last message (``Press Ctrl+c to exit `odo dev` and delete resources from the cluster``) remains green, which might cause visual confusion with the list of ports forwarded.

image

**Comment by rm3l** [comment]
> One last thing I noticed after testing your branch. The color of the last message (`` Press Ctrl+c to exit `odo dev` and delete resources from the cluster ``) remains green, which might cause visual confusion with the list of ports forwarded.
> 
> image

Weird, I launched the same command again, and now the message shows up as white 🤔

image

File: pkg/watch/interface.go
Comment (dharmit) [reply pkg/watch/interface.go]: Makes sense. Removed.

File: pkg/odo/cli/dev/dev.go
Comment (rm3l) [inline pkg/odo/cli/dev/dev.go:239]: I'm not an expert with channels and goroutines, but I feel like attempting to read from the Done channel with `<-o.ctx.Done()` is already a blocking call, which won't continue until the `o.ctx` is done.
So I think this infinite `for` loop is not necessary, especially since there is no other `select` cases there, no?
I quickly tested without the `for` loop and it worked as expected. What do you think?

Comment (dharmit) [reply pkg/odo/cli/dev/dev.go:239]: That's right. I had initially created a `select` with a single case of listening on `o.ctx`. I removed the `select`, but didn't realise that `for` could be removed as well. Thanks for pointing. 👍🏾

Comment (dharmit) [reply pkg/odo/cli/dev/dev.go]: Done. Thanks for the catch.

**Review by rm3l — APPROVED** [review]

**Comment by dharmit** [comment]
Windows test failure:
```
------------------------------
[7] + Failure in Spec Setup (BeforeEach) [362.751 seconds]
[7] odo dev command tests
[7] C:/Users/Administrator.ANSIBLE-TEST-VS/637/tests/integration/devfile/cmd_dev_test.go:23
[7]   when a component is bootstrapped and pushed
[7]   C:/Users/Administrator.ANSIBLE-TEST-VS/637/tests/integration/devfile/cmd_dev_test.go:53
[7]     when odo is executed with --no-watch flag
[7]     C:/Users/Administrator.ANSIBLE-TEST-VS/637/tests/integration/devfile/cmd_dev_test.go:334
[7]       when a file in component directory is modified [BeforeEach]
[7]       C:/Users/Administrator.ANSIBLE-TEST-VS/637/tests/integration/devfile/cmd_dev_test.go:347
[7]         should not trigger a push
[7]         C:/Users/Administrator.ANSIBLE-TEST-VS/637/tests/integration/devfile/cmd_dev_test.go:348
[7] 
[7]         Timed out after 360.009s.
[7]         Expected
[7]             <string>: 
[7]         to contain substring
[7]             <string>: Press Ctrl+c to exit `odo dev` and delete resources from the cluster
[7] 
```

**Comment by dharmit** [comment]
> Seeing a panic in integration tests running on Windows.

Tried `odo dev --no-watch` manually on Windows, it works fine. Not sure what's the problem. 😕

**Comment by dharmit** [comment]
Weirder than before. It can't find `--no-watch` flag! 😕 
```
[14] [odo] Your new component 'qaymxq' is ready in the current directory.
[14] [odo] To start editing your component, use 'odo dev' and open this folder in your favorite IDE.
[14] [odo] Changes will be directly reflected on the cluster.
[14] Running odo.exe with args [odo dev --random-ports --no-watch]
[14] [odo] Error: unknown flag: --no-watch
[14] [odo] Usage:
[14] [odo]   odo dev [flags]
[14] [odo] 
[14] [odo] Examples:
[14] [odo]   # Deploy component to the development cluster
[14] [odo]   odo dev
[14] [odo] 
[14] [odo] Flags:
[14] [odo]   -h, --help           Help for dev
[14] [odo]   -f, --random-ports   Assign random ports to redirected ports
[14] [odo] 
[14] [odo] Additional Flags:
[14] [odo]       --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
[14] [odo]   -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
[14] [odo]       --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
[14] [odo] 
[14] [odo]  X  unknown flag: --no-watch
```

**Comment by dharmit** [comment]
@rm3l @kadel all three of us have played with this PR on the Windows, and it's working fine/as expected. Can you folks help merge this PR?

**Comment by rm3l** [comment]
> Weirder than before. It can't find `--no-watch` flag! confused
> 
> ```
> [14] [odo] Your new component 'qaymxq' is ready in the current directory.
> [14] [odo] To start editing your component, use 'odo dev' and open this folder in your favorite IDE.
> [14] [odo] Changes will be directly reflected on the cluster.
> [14] Running odo.exe with args [odo dev --random-ports --no-watch]
> [14] [odo] Error: unknown flag: --no-watch
> [14] [odo] Usage:
> [14] [odo]   odo dev [flags]
> [14] [odo] 
> [14] [odo] Examples:
> [14] [odo]   # Deploy component to the development cluster
> [14] [odo]   odo dev
> [14] [odo] 
> [14] [odo] Flags:
> [14] [odo]   -h, --help           Help for dev
> [14] [odo]   -f, --random-ports   Assign random ports to redirected ports
> [14] [odo] 
> [14] [odo] Additional Flags:
> [14] [odo]       --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
> [14] [odo]   -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
> [14] [odo]       --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
> [14] [odo] 
> [14] [odo]  X  unknown flag: --no-watch
> ```

Indeed, this is strange. As you said, it looks like a different binary was used. I can see the version is `v2.5.0` in the CI job logs, while it should now be `v3.0.0-alpha1` (since 946385b and #5677, and given that you rebased your branch onto `main`):

```
[14] [odo] I0421 03:47:24.159538   10020 implem.go:103] The path for preference file is C:\Users\Administrator.ANSIBLE-TEST-VS\AppData\Local\Temp\238738076\preference.yaml
[14] [odo]   __
[14] [odo]  /  \__     Initializing new component
[14] [odo]  \__/  \    
[14] [odo]  /  \__/    odo version: v2.5.0
[14] [odo]  \__/
[14] [odo] 
[14] [odo] Interactive mode enabled, please answer the following questions:
[14] [odo]  -  Copying devfile from "C:\\Users\\Administrator.ANSIBLE-TEST-VS\\651\\tests\\examples\\source\\devfiles\\nodejs\\devfile.yaml"  ...
[14] [odo]
```

LGTM after retesting your changes, but maybe we should create a separate issue for this.

/lgtm
/approve

**Comment by rm3l** [comment]
Don't know why, but a lot of unit tests CI jobs are currently Queued in IBM Cloud. And this is blocking this PR from getting merged. Is there any way to dequeue them?

**Comment by dharmit** [comment]
Overriding because unit tests already succeeded using prow

/override Unit Tests

**Comment by dharmit** [comment]
> Don't know why, but a lot of unit tests CI jobs are currently Queued in IBM Cloud. And this is blocking this PR from getting merged. Is there any way to dequeue them?

I have overridden unit tests since the counterparts are already passing on prow. And for Windows, well, overrode that as well. Let's see.

**Pull request merged**
